### PR TITLE
Add WER metric

### LINKED ISSUE #28: Add WER metric

### PR DESCRIPTION
**Pull request by nikolasavic3**
Add Word Error Rate (WER) metric 
Fixes #28

### CODE REVIEW

File: src/metrax/metrics.py [hunk 1]
```diff
@@ -795,3 +795,140 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
...
+    m, n = len(prediction), len(reference)
+
+    # Handle edge cases
+    if m == 0:
+      return n
+    if n == 0:
+      return m
+
+    # Create distance matrix
+    distance_matrix = [[0 for _ in range(n+1)] for _ in range(m+1)]
```
Comment (nikolasavic3) [inline src/metrax/metrics.py:901]: I implemented the Levenshtein distance calculation using pure Python rather than JAX operations because the distance calculation involves updating the matrix in-place, and if we use JAX, that would create a new copy of the array for every update. I also believe that typical WER use cases have relatively short sequences, so the performance difference is negligible. Only the final results are converted to JAX arrays, which follows the pattern used in other metrics in the library.

File: src/metrax/metrics_test.py [hunk 2]
```diff
@@ -320,6 +320,148 @@ def test_rsquared(self, y_true, y_pred, sample_weights):
...
+          None,
+      ),
+      (
+          'weighted',
+          np.random.randint(10, size=[2, 5, 10]),
+          np.random.uniform(size=(2, 5, 10, 20)),
+          np.random.randint(2, size=(2, 5, 10)).astype(np.float32),
+      ),
+  )
+  def test_perplexity(self, y_true, y_pred, sample_weights):
```
Comment (jshin1394) [inline src/metrax/metrics_test.py:337]: not sure why this shows up as part of this PR. This code already exists...?

File: src/metrax/metrics_test.py [hunk 3]
```diff
@@ -320,6 +320,148 @@ def test_rsquared(self, y_true, y_pred, sample_weights):
...
+
+    np.testing.assert_allclose(
+        merged_metric.compute(),
+        jnp.array(3/18, dtype=jnp.float32),
+        rtol=1e-05,
+        atol=1e-05,
+    )
+
 
 if __name__ == '__main__':
```
Comment (jshin1394) [inline src/metrax/metrics_test.py:466]: could you try syncing your workspace seems like some of more recent updates are not reflected.

File: src/metrax/metrics_test.py [hunk 4]
```diff
@@ -320,6 +320,148 @@ def test_rsquared(self, y_true, y_pred, sample_weights):
...
+
+    expected = keras_metric.result()
+    np.testing.assert_allclose(
+        metrax_metric.compute(),
+        expected,
+        rtol=1e-05,
+        atol=1e-05,
+    )
+
+  def test_wer_empty(self):
```
Comment (jshin1394) [inline src/metrax/metrics_test.py:360]: let's move all *_empty tests to the beginning of the file.

File: src/metrax/metrics_test.py [hunk 5]
```diff
@@ -320,6 +320,148 @@ def test_rsquared(self, y_true, y_pred, sample_weights):
...
+    for pred, ref in zip(predictions, references):
+      update = metrax.WER.from_model_output(
+          predictions=[pred],
+          references=[ref],
+      )
+      metric = update if metric is None else metric.merge(update)
+
+    np.testing.assert_allclose(
+        metric.compute(),
+        jnp.array(3/18, dtype=jnp.float32),
```
Comment (jshin1394) [inline src/metrax/metrics_test.py:395]: can we test against the result gained from using a different known library? for instance, we can compare against torchmetrics https://pytorch.org/torcheval/main/generated/torcheval.metrics.WordErrorRate.html

File: src/metrax/metrics.py [hunk 6]
```diff
@@ -536,4 +536,141 @@ def compute(self) -> jax.Array:
...
+      raise ValueError(
+          f'Length mismatch: predictions has {len(predictions)} items, '
+          f'but references has {len(references)} items'
+      )
+
+    # Determine if inputs are strings that need tokenization or already tokenized
+    total_edit_distance = 0
+    total_reference_length = 0
+
+    for pred, ref in zip(predictions, references):
```
Comment (jshin1394) [inline src/metrax/metrics.py:605]: should we add a TODO to use a more involved tokenizer in the future?

**Comment by jshin1394** [comment]
Thank you Nikola for your PR! Left some comments. Feel free to sync the workspace as well! :)

File: src/metrax/metrics.py (hunk 1 above)
Comment (jshin1394) [reply src/metrax/metrics.py:898]: sounds good to me, thank you!

File: src/metrax/metrics.py [hunk 7]
```diff
@@ -795,3 +795,137 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
 
   def compute(self) -> jax.Array:
     return jnp.exp(self.aggregate_crossentropy / self.num_samples)
+
+
+@flax.struct.dataclass
+class WER(clu_metrics.Metric):
```
Comment (jshin1394) [inline src/metrax/metrics.py:801]: should your metric inherit from clu_metrics.Average?

Comment (nikolasavic3) [reply src/metrax/metrics.py:801]: What do you think about it?

File: src/metrax/metrics_test.py (hunk 5 above)
Comment (nikolasavic3) [reply src/metrax/metrics_test.py:488]: Yes, that's a good idea. Maybe in the future, we could open a PR for keras_hub to implement the WER metric there so we have less imports? It already has ROUGE and BLEU. Or maybe we could use a simpler library like https://github.com/jitsi/jiwer.

File: src/metrax/metrics.py (hunk 6 above)
Comment (nikolasavic3) [reply src/metrax/metrics.py:863]: That's a great point! To keep things in line with the single responsibility principle, I think we should remove the tokenization logic entirely and just accept pre-tokenized lists of strings as input, leaving the tokenizer implementation to the one who calls the metric.

**Comment by nikolasavic3** [comment]
Thank you Jiwon for reviewing my PR!

File: src/metrax/metrics.py (hunk 7 above)
Comment (jshin1394) [reply src/metrax/metrics.py:801]: Hi Nikola! We definitely want to use these metrics in distributed settings. However, I wanted to point out that clu_metrics.Average also inherits from the clu_metrics.Metric class. Therefore, inheriting from clu_metrics.Average should be just as upgradable as clu_metrics.Metric. Please let me know if I am missing anything. Otherwise, +1 to using Average as base class! :)

**Comment by jshin1394** [comment]
Thank you Nikola! LGTM given all the comments are resolved.

**Comment by jshin1394** [comment]
Hi nikola, a bunch of PRs were submitted to split the metrics.py file into multiple modules. I think the WER class can be placed under the new `nlp_metrics.py`module.

File: src/metrax/metrics_test.py (hunk 5 above)
Comment (nikolasavic3) [reply src/metrax/metrics_test.py]: the test now use [keras_nlp.metrics.EditDistance](https://keras.io/api/keras_nlp/metrics/edit_distance/) with `normalize=True` which is the same as computing WER, as mentioned in the documentation.

File: src/metrax/metrics.py (hunk 7 above)
Comment (nikolasavic3) [reply src/metrax/metrics.py]: You’re right, the class now inherits from Average.

File: src/metrax/nlp_metrics.py [hunk 8]
```diff
@@ -114,3 +117,126 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
...
+      total_reference_length: Sum of reference lengths across all samples.
+  """
+  @property
+  def total_edit_distance(self):
+    return self.total
+  @property
+  def total_reference_length(self):
+    return self.count
+
+  @classmethod
```
Comment (jshin1394) [inline src/metrax/nlp_metrics.py:153]: no need to override this function as it already exists in `Average`

File: src/metrax/nlp_metrics.py [hunk 9]
```diff
@@ -114,3 +117,126 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
...
+
+        distance_matrix[i][j] = min(
+            distance_matrix[i-1][j] + 1,      # deletion
+            distance_matrix[i][j-1] + 1,      # insertion
+            distance_matrix[i-1][j-1] + cost  # substitution
+        )
+
+    return distance_matrix[m][n]
+
+  def merge(self, other: 'WER') -> 'WER':
```
Comment (jshin1394) [inline src/metrax/nlp_metrics.py:233]: no need to override this function anymore as it exists in `Average`

File: src/metrax/metrics.py (hunk 6 above)
Comment (jshin1394) [reply src/metrax/metrics.py]: Hi Nikola! I agree that single responsibility principle is important. However, considering that it is more common for these types of metrics to take in a whole sentence, (for instance, this [torchMetrics WER](https://pytorch.org/torcheval/main/generated/torcheval.metrics.WordErrorRate.html)) let's keep the implementation such that we can take in sentences as input.

File: src/metrax/metrics_test.py (hunk 5 above)
Comment (jshin1394) [reply src/metrax/metrics_test.py]: perfect. Thank you so much!

File: src/metrax/nlp_metrics.py [hunk 10]
```diff
@@ -114,3 +117,126 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
...
+    return type(self)(
+        total=self.total + other.total,
+        count=self.count + other.count,
+    )
+
+  def compute(self) -> jax.Array:
```
Comment (jshin1394) [inline src/metrax/nlp_metrics.py:239]: no need to override this function as it exists in `Average`

File: src/metrax/nlp_metrics.py [hunk 11]
```diff
@@ -114,3 +117,126 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
...
+
+  A lower WER indicates better performance, with 0 being perfect.
+
+  This implementation only accepts pre-tokenized inputs (lists of tokens).
+
+  Attributes:
+      total_edit_distance: Sum of edit distances across all samples.
+      total_reference_length: Sum of reference lengths across all samples.
+  """
+  @property
```
Comment (jshin1394) [inline src/metrax/nlp_metrics.py:146]: these can be removed as well since

**Comment by jshin1394** [comment]
Thank you so much nikola! Almost there :) I left some comments so that we can 1) remove functions that already exist as part of [`Average`](https://github.com/google/CommonLoopUtils/blob/main/clu/metrics.py#L770) and 2) the metric can take in sentence as input following other popular metric libraries such as [torchMetrics](https://pytorch.org/torcheval/main/generated/torcheval.metrics.WordErrorRate.html).

Thank you again Nikola!

**Comment by nikolasavic3** [comment]
Thank you for the detailed feedback, Jiwon! I've removed all the redundant function overrides (merge, compute, total_reference_length) and updated the implementation to accept untokenized strings.

**Comment by jshin1394** [comment]
Thank you so much Nikola! We really appreciate your contribution to the Metrax codebase! :)
